File: tell-dont-ask-kata-ruby/use_cases/OrderApprovalUseCase.rb
require_relative "RejectedOrderCannotBeApprovedError"
require_relative "ShippedOrdersCannotBeChangedError"
require_relative "ApprovedOrderCannotBeRejectedError"

class OrderApprovalUseCase

    attr_accessor :order_database

    def initialize(order_database:)
        @order_database = order_database
    end

    def run(request:)
        order = @order_database.find_by_id(:order_id => request.order_id)

        if (order.status == OrderStatus::SHIPPED)
            raise ShippedOrdersCannotBeChangedError.new
        end

        if (request.is_approved? && order.status == OrderStatus::REJECTED)
            raise RejectedOrderCannotBeApprovedError.new
        end

        if (!request.is_approved? && order.status == OrderStatus::APPROVED)
            raise ApprovedOrderCannotBeRejectedError.new
        end

        order.status = request.is_approved? ? OrderStatus::APPROVED : OrderStatus::REJECTED
        @order_database.save(:order => order)
    end
end

File: tell-dont-ask-kata-csharp/TellDontAskKata/useCase/OrderApprovalUseCase.cs
public class OrderApprovalUseCase
{
    private readonly OrderRepository orderRepository;

    public OrderApprovalUseCase(OrderRepository orderRepository)
    {
        this.orderRepository = orderRepository;
    }

    public void Run(OrderApprovalRequest request)
    {
        Order order = orderRepository.GetById(request.GetOrderId());

        if (order.GetStatus().Equals(OrderStatus.SHIPPED))
        {
            throw new ShippedOrdersCannotBeChangedException();
        }

        if (request.IsApproved() && order.GetStatus().Equals(OrderStatus.REJECTED))
        {
            throw new RejectedOrderCannotBeApprovedException();
        }

        if (!request.IsApproved() && order.GetStatus().Equals(OrderStatus.APPROVED))
        {
            throw new ApprovedOrderCannotBeRejectedException();
        }

        order.SetStatus(request.IsApproved() ? OrderStatus.APPROVED : OrderStatus.REJECTED);
        orderRepository.Save(order);
    }
}